### PR TITLE
feat: highlight search matches in results

### PR DESCRIPTION
**Pull request by y-lohse**

### CODE REVIEW

File: src/components/SearchBar.jsx [hunk 1]
```diff
@@ -7,13 +7,68 @@ const INTENT_VERB = 'OPEN'
...
+    })
+
+    currentIndex += resultTerm.length + spacerLength
+  })
+
+  // matchedIntervals can overlap, so we merge them.
+  // - sort the intervals by starting index
+  // - add the first interval to the stack
+  // - for every interval,
+  // - - add it to the stck if it doesn't overlap with the stack top
```
Comment (CPatchane) [inline src/components/SearchBar.jsx]: s/stck/stack ?

Comment (y-lohse) [reply src/components/SearchBar.jsx]: 👍

File: src/components/SearchBar.jsx [hunk 2]
```diff
@@ -7,13 +7,68 @@ const INTENT_VERB = 'OPEN'
...
+  const normalizedQueryTerms = normalizeString(query)
+  const normalizedResultTerms = normalizeString(searchResult)
+
+  const matchedIntervals = []
+  const spacerLength = 1
+  let currentIndex = 0
+
+  normalizedResultTerms.forEach(resultTerm => {
+    normalizedQueryTerms.forEach(queryTerm => {
+      const index = resultTerm.indexOf(queryTerm)
```
Comment (enguerran) [inline src/components/SearchBar.jsx]: I would prefer a `regex` search

File: src/components/SearchBar.jsx [hunk 3]
```diff
@@ -7,13 +7,68 @@ const INTENT_VERB = 'OPEN'
...
+      if (index >= 0) {
+        matchedIntervals.push({
```
Comment (enguerran) [inline src/components/SearchBar.jsx]: don'tlike `forEach` with side effects. Can't we have a `map`?

File: src/components/SearchBar.jsx [hunk 4]
```diff
@@ -7,13 +7,68 @@ const INTENT_VERB = 'OPEN'
...
+    currentIndex += resultTerm.length + spacerLength
+  })
+
+  // matchedIntervals can overlap, so we merge them.
+  // - sort the intervals by starting index
+  // - add the first interval to the stack
+  // - for every interval,
+  // - - add it to the stck if it doesn't overlap with the stack top
+  // - - or extend the stack top if the start overlaps and the new interval's top is bigger
+  const mergedIntervals = matchedIntervals.sort((intervalA, intervalB) => intervalA.from > intervalB.from).reduce((computedIntervals, newInterval) => {
```
Comment (enguerran) [inline src/components/SearchBar.jsx]: I guess the algorithme is not mastered as it needs some comments.

File: src/components/SearchBar.jsx (hunk 2 above)
Comment (y-lohse) [reply src/components/SearchBar.jsx]: That's subjective though. I prefer `indexOf` for searches that don't involve pattern matching.

File: src/components/SearchBar.jsx (hunk 3 above)
Comment (y-lohse) [reply src/components/SearchBar.jsx]: No, precisely because of the "side effects", since they aren't side effects at all. They aren't a byproduct of the `forEach`, they are the reason for it.
More extensively, I'm creating new information based on 2 existing arrays, not reducing or transforming the existing data. The array I'm building could have less, the same, or more elements than any of the base arrays. And the data it contains is not a transformation of the values of the existing arrays either.

File: src/components/SearchBar.jsx (hunk 2 above)
Comment (enguerran) [reply src/components/SearchBar.jsx]: did you benchmark it, I guess regex are way faster than indexOf: https://jsperf.com/regexp-indexof-perf

Comment (y-lohse) [reply src/components/SearchBar.jsx]: From what I've seen, it depends on the browser and on the test case.
In the test case you've linked, only `test` is used, whereas here we want the index of the match, so `String.search` would be more appropriate. It's also doing the tests on node class names, which we're really not doing here. [This is another benchmark](https://jsperf.com/exec-vs-match-vs-test-vs-search/5) that I just ran, and `indexOf` just blows everything off the water (849,050,703 ops vs 17,738,919 for `search`and `20,586,521` for `test`).

File: src/components/SearchBar.jsx (hunk 4 above)
Comment (y-lohse) [reply src/components/SearchBar.jsx]: I found the algorithm not so trivial, enough to warrant a quick explanation in plain english.

File: src/components/SearchBar.jsx (hunk 2 above)
Comment (enguerran) [reply src/components/SearchBar.jsx]: ok

File: src/components/SearchBar.jsx (hunk 3 above)
Comment (enguerran) [reply src/components/SearchBar.jsx]: When you say:
> No, precisely because of the "side effects", since they aren't side effects at all. 

You mean the code could be refactored like this:
```js
normalizedResultTerms.forEach(newFunction1())

function newFunction1() {
  return resultTerm => {
    normalizedQueryTerms.forEach(newFunction(resultTerm));
    currentIndex += resultTerm.length + spacerLength;
  };
}

function newFunction(resultTerm) {
  return queryTerm => {
    const index = resultTerm.indexOf(queryTerm)
    if (index >= 0) {
      matchedIntervals.push({
        from: currentIndex + index,
        to: currentIndex + index + queryTerm.length
      })
    }
  }
}
```
with `newFunction` and newFunction1` as [pure functions](https://en.wikipedia.org/wiki/Pure_function) ?

I mean: you need to be opinionated to write some code. But the reason you give to justify the point is not good, nor real.
Nevertheless you can keep this code instead of writing a real pure functional version. There is no problem to have this code on the codebase. It is just a matter of readability and maintainability.
